mmv1/products/kms/terraform.yaml

```diff
@@ -98,6 +98,24 @@ overrides: !ruby/object:Overrides::ResourceOverrides
       decoder: templates/terraform/decoders/kms.go.erb
       encoder: templates/terraform/encoders/kms_crypto_key.go.erb
       update_encoder: templates/terraform/update_encoder/kms_crypto_key.go.erb
+  CryptoKeyVersion: !ruby/object:Overrides::Terraform::ResourceOverride
+    description: |
+      {{description}}
+
+      Destroying a cryptoKeyVersion will not delete the resource from the project.
+    id_format: "{{name}}"
+    import_format: ["{{name}}"]
+    examples:
+      - !ruby/object:Provider::Terraform::Examples
+        name: "kms_crypto_key_version_basic"
+        primary_resource_id: "example-key"
+        skip_test: true
+    properties:
+      state: !ruby/object:Overrides::Terraform::PropertyOverride
+        default_from_api: true
+    custom_code: !ruby/object:Provider::Terraform::CustomCode
+      custom_delete: templates/terraform/custom_delete/kms_crypto_key_version.erb
```

Comment:
Currently the custom_delete code has the following logic:

1. Try to parse the id; if that fails return the error
2. If the id is parseable but nil, don't set the id, but return nil (will not remove the resource from state)
   - Note: this is not reachable code, because the id can only be nil if there's an error
4. Otherwise, set the id to "" (which means 'delete from state') and return nil

The logic we probably want to have looks more like:

1. Try to parse the id; if that fails return the error
2. try to call the `destroy` method for the resource with the given id. This will be similar to the logic in `clearCryptoKeyVersions` in `kms_utils.go`.
3. Handle the response with `handleNotFoundError`

This will be similar to what [`resourceGoogleProjectDelete`](https://github.com/hashicorp/terraform-provider-google/blob/3069f7581c887e9783e0026a210378d883fc3131/google/resource_google_project.go#L446) does (except without `skip_delete` support.) 

I'm not sure how the API will behave if the resource is already scheduled for deletion. If it just returns a not found error, then this should be enough. Otherwise you might need more complicated logic. 🍀

Reply:
Thanks for the detailed explanation. I will take a look for "try to call the destroy method". I'm slightly concerned with the resource is already scheduled for deletion error as, during deletion of crypto_key, it also deletes the corresponding version. But beforehand I should start with the destroy method.

Reply:
Still working on it, the deletion URL still has some issue, will investigate on it.

Reply:
Attempted the following on the latest commit:
(1) CryptoKeyVersion: Added a custom delete function, calling for destroy.
(2) CryptoKey: Do not delete versions at deletion.

It turn out that we get:
**github.com/hashicorp/terraform-provider-google-beta/google-beta.deleteCryptoKeyVersions(0xc015121f90, 0x88?, {0xc0208a62a0?, 0x6a?}, 0xc00efc8400)
	/go/src/github.com/hashicorp/terraform-provider-google-beta/google-beta/kms_utils.go:249 +0x223**

Locally, I attempted the following:
(1) CryptoKeyVersion: Added a custom delete function, calling for destroy.
(2) Keep CryptoKey deleting the versions
Ended up into **"message": "The request cannot be fulfilled. Resource projects/tf-test-1111180836049777411/locations/us-central1/keyRings/tf-test-071bb6qyfe/cryptoKeys/tf-test-9gmcgjtnnf/cryptoKeyVersions/1 has value DESTROY_SCHEDULED in field crypto_key_version.state."**

(3) I added a check to handle "DESTROY_SCHEDULED" on step 2, then the very first error appears again.

Reply:
so, it looks like the current commit has the following full error:

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0xc0 pc=0x1ea19c3]

goroutine 217139 [running]:
github.com/hashicorp/terraform-provider-google-beta/google-beta.deleteCryptoKeyVersions(0xc015121f90, 0x88?, {0xc0208a62a0?, 0x6a?}, 0xc00efc8400)
	/go/src/github.com/hashicorp/terraform-provider-google-beta/google-beta/kms_utils.go:249 +0x223
```

The referenced line is: https://github.com/modular-magician/terraform-provider-google-beta/blob/32810c33a6bef8bdcf84fa1d3c43806ebbd0c7dd/google-beta/kms_utils.go#L249

The panic indicates that `response` is nil, which probably means that the `err` is not nil. Possibly swapping the `if err != nil` block to come first would resolved this error & get the behavior you want?

Reply:
Sounds good, let me try that.

Reply:
Looks like it works, but crypto_key tests are broken because the versions are expected to be deleted.

I'm trying to add the exception handler on clearCryptoKeyVersion (called when deleting crypto_key) and see if that works. So far I'm keep getting **"message": "The request cannot be fulfilled. Resource projects/tf-test-1111180836049777411/locations/us-central1/keyRings/tf-test-071bb6qyfe/cryptoKeys/tf-test-9gmcgjtnnf/cryptoKeyVersions/1 has value DESTROY_SCHEDULED in field crypto_key_version.state."**

Alternatively I can add cryptoKeyVersion to all existing cryptoKey tests.